src/statistical/unigram.jl

```diff
@@ -0,0 +1,221 @@
+    logprob = Float64[]
+    for i in 1:length(vocab)
+        vocab1 = vocabnew[i][1]
+        vocab1 = replace(vocab1, "▁"=>"_")
+        push!(voc, vocab1)
+        logp = vocabnew[i][2]
+        logp =  parse(Float64, logp)
+        push!(logprob, logp)    
+    end
+    unk_id = findall(x->x==unk_token, voc)
+    length(unk_id) == 0 && throw(UndefVarError(:unk_token)) 
+    spm = SentencePieceModel(voc, logprob, unk_id[1])
+    return spm
+end
+
+# to get index of given string
+function getindex(sp::SentencePieceModel, text)
+    id_list = findall(x->x==text, sp.vocab)
+    length(id_list) == 0 && return sp.unk_id #unk token index 
+    return id_list[1]
+end
+
+"""
+struct Nodes 
+    text::String
+    score::Float32
+    index::Int64
+    start::Int
+    en::Int
+end
+Utility structure, To hold the results of the `forward pass` (the forward Viterbi lattice)
+hold the token token string, score, vocabulary index, start and end character position   
+"""
+struct Nodes 
+    text::String
+    score::Float32
+    index::Int64
+    start::Int
+    en::Int
+end
+
+"""
+    decode_forward(sp::SentencePieceModel,text::String)
+Perform `forward pass` (the forward Viterbi lattice) operation detail can be found [here](https://tejasvaidhyadev.github.io/blog/Sentencepiece).
+Return all output, as an Array{String,1}
+# Example
+```julia-repl
+julia> seq = ["To","be","or","not"]
+julia> node = WordTokenizers.decode_forward(spm, "I love julia language")
+21-element Array{WordTokenizers.Nodes,1}:
+ WordTokenizers.Nodes("I", -Inf32, 1, 1, 1)
+ WordTokenizers.Nodes(" ", -Inf32, 1, 2, 2)
+ WordTokenizers.Nodes("l", -Inf32, 1, 3, 3)
+ WordTokenizers.Nodes("lo", -9.56041f0, 1416, 3, 4)
+ WordTokenizers.Nodes("lov", -11.0086f0, 5943, 3, 5)
+ WordTokenizers.Nodes("love", -10.7128f0, 4584, 3, 6)
+ WordTokenizers.Nodes(" ", -Inf32, 1, 7, 7)
+ WordTokenizers.Nodes("j", -Inf32, 1, 8, 8)
+ WordTokenizers.Nodes("ju", -9.96107f0, 2143, 8, 9)
+ WordTokenizers.Nodes("ul", -19.4301f0, 1288, 9, 10)
+ WordTokenizers.Nodes("uli", -21.02407f0, 6244, 9, 11)
+ WordTokenizers.Nodes("ulia", -22.49547f0, 19590, 9, 12)
+ WordTokenizers.Nodes(" ", -Inf32, 1, 13, 13)
+ WordTokenizers.Nodes("l", -Inf32, 1, 14, 14)
+ WordTokenizers.Nodes("la", -8.6488f0, 532, 14, 15)
+ WordTokenizers.Nodes("lan", -9.78918f0, 1805, 14, 16)
+ WordTokenizers.Nodes("lang", -11.6118f0, 9950, 14, 17)
+ WordTokenizers.Nodes("gu", -21.9203f0, 3074, 17, 18)
+ WordTokenizers.Nodes("gua", -23.776f0, 15259, 17, 19)
+ WordTokenizers.Nodes("ag", -34.1531f0, 3303, 19, 20)
+ WordTokenizers.Nodes("language", -11.1965f0, 7021, 14, 21)
+``` 
+"""
+function decode_forward(sp::SentencePieceModel, text::String)
+    results = Array{Nodes, 1}(undef, length(text))
+    scores = fill(-Inf, length(text))
+    scores[1] =0
+    for char_end in 1:length(text)
+        for char_start in 1:char_end
+            if text[char_start:char_end] in sp.vocab
+                subtokenid = getindex(sp, text[char_start:char_end])[1]
```

Comment:
if we use `substring` then we can avoid allocating a copy of the string, so it should be faster.
```suggestion
            subtoken = substring(text, char_start:char_end)
            if subtoken in sp.vocab
                subtokenid = sp[subtoken][1]
```